src/v/storage/offset_to_filepos.cc

```diff
@@ -85,7 +85,22 @@ offset_to_filepos_consumer::type offset_to_filepos_consumer::end_of_stream() {
 
 } // namespace internal
 
-ss::future<offset_to_file_pos_result> convert_begin_offset_to_file_pos(
+std::ostream&
+operator<<(std::ostream& os, convert_to_file_pos_outcome conv_res) {
+    switch (conv_res) {
+    case success:
+        os << "success";
+        break;
+    case offset_not_in_segment:
+        os << "offset_not_in_segment";
+        break;
+    }
+
+    return os;
+}
+
+ss::future<checked<offset_to_file_pos_result, convert_to_file_pos_outcome>>
```

Comment:
i would recommend not using `checked`.

in this case `convert_to_file_pos_outcome` is either success or a known error, so you could already represent this as an `std::optional<pos_result>` or `result<std::optional<pos_result>>` if you needed to communicate additional errors.

if `convert_to_file_pos_outcome` had more than 1 non-success error values (which maybe that is the intention in the future), then it'd likely be better to create proper std::error code integration so that it could be used with `result<T>`.

Reply:
I started off thinking that there may be more than one error outcomes so the enum is a placeholder, this function throws an exception [here](https://github.com/redpanda-data/redpanda/blob/6574ef18f9ef043194e9d2775f4f1fa641a26a5d/src/v/storage/offset_to_filepos.cc#L142) which I thought could be an error rather than an exception. I still have to discuss this with Evgeny if not creating an upload candidate is fine in that case, or if we must throw an exception.

In either case I will move the code to `result<T>` and using standard error code, thanks for the suggestion.

Reply:
I have used `std::errc::result_out_of_range` and `result<offset_to_file_pos_result>` for now, will look into using a set of custom error codes if we decide to remove the exception thrown in these functions

Reply:
> this function throws an exception [here](https://github.com/redpanda-data/redpanda/blob/6574ef18f9ef043194e9d2775f4f1fa641a26a5d/src/v/storage/offset_to_filepos.cc#L142) which I thought could be an error rather than an exception.

i suspect @mmaslankaprv has enough experience with result<T> in raft to have some good advice. imo the important thing is to be consistent in a sub-system. for example, if result<T> is what is used for errors, then a ss::future<result<T>> could still contain an exception, but the goal would be to minimize application specific exceptions, and catch stuff like bad_alloc etc... all up at a higher level.